Add: 회원가입 로직 구현

minju1009 opened the pull request:
### PR Type
- [ ] 버그 수정
- [x] 신규 기능
- [ ] 기타 (기타인 경우 내용 입력 :  ) 

### 해당 PR의 목적
- 회원가입 기능 추가

### 해당 PR 사유
- 회원가입 기능 구현에 대한 피드백 요청

### 리뷰어가 중점적으로 봤으면 하는 부분
1. `useEffect`와 `setTimout`을 사용해 debounce통신을 구현하였는데 잘 구현되었는지 피드백 요청드립니다!

minju1009 committed:
Add: 회원가입 로직 구현

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+    return () => clearTimeout(debounce);
+  }, [email]);
+
+  const checkEmailValidation = (emailAddress: string) => {
```
`valiateEmail()`정도로 짧게 표현할 수 있을 것 같습니다.

minju1009 replied:
네, 변수명 수정하였습니다 :)

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+    password: '',
+    passwordCheck: '',
+  });
+  const [checkEmailUniqueness, setCheckEmailUniqueness] = useState('');
```
`isUniqueEmail`정도로 짧게 표현할 수 있을 것 같습니다.

minju1009 replied:
변수명 짧고 간결하게 표현하도록 하겠습니다! 해당 부분 수정하였습니다!

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+
+  useEffect(() => {
+    const fetchData = () => {
+      fetch(`API.signup?email=${email}`)
```
`API.signup`이 string으로 되어있어 정상적으로 작동하지 않을 것 같습니다.

minju1009 replied:
템플릿 리터럴`${API.signup}?email=${email}`로 변경하였습니다.

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+  };
+
+  useEffect(() => {
+    const fetchData = () => {
```
`fetchData`라는 변수명을 조금 더 구체적으로 작성해주세요, 
`state`값에 포함된 `unique`라는 단어도 좋고, `fetchData()`는 너무 광범위합니다.

minju1009 replied:
`checkIsUnique`로 수정해주었습니다!

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+    return emailRegex.test(emailAddress);
+  };
+
+  const emailErrorMsg = checkEmailValidation(email)
```
함수가 아닌 변수를 이렇게 설정하면, state 업데이트가 일어날 때마다 `checkEmailValidation()`이라는 함수를 호출하게 됩니다. 
email인풋이 바뀔때만 호출하거나, `useMemo()`를 사용해서 dependency에 따라서만 호출될 수 있도록 수정해주세요

minju1009 replied:
`useMemo()`를 사용하여 `email`에 변경이 있을 때에만 호출되도록 바꿔주었습니다.

minju1009 replied:
`useMemo()`로 감싸서 email이 변경될 때에만 호출되도록 하였습니다.

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+    return pwRegex.test(pw) && pw.length > 0;
+  };
+
+  const validator = {
```
이 변수에도 state 업데이트가 일어날 때마다 값들이 할당되게 됩니다. 변수의 선언은 `useState`, `useRef`, `useMemo`를 사용해주세요

minju1009 replied:
네 알겠습니다! 해당 변수를 없애고, 유저가 입력하는 모든 값에 대한 validation은 useEffect를 통해 처리해주었습니다.

minju1009 replied:
네! 알겠습니다. 해당 부분 useState 사용으로 변경하였습니다.

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+    passwordCheck: password === passwordCheck,
+  };
+
+  const isEverythingValid =
```
여기도 마찬가지로 계속해서 할당이 일어나게 됩니다

minju1009 replied:
유저가 입력하는 값에 대한 할당만 일어나도록 `useMemo` 사용해주었습니다.

minju1009 replied:
수정완료하였습니다!

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+          />
+          {password.length > 0 && (
+            <ErrorMsg>
+              {validator.password
```
`ErrorMessage`를 보여주는 부분은 삼항연산자 보다는, 에러가 있을때만 표시될 수 있도록 하는 편이 좋습니다.

minju1009 replied:
넵! 삼항연산자가 아닌 && 연산자를 통해 다시 표현해주었습니다.

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+
+  const {lastName, firstName, email, password, passwordCheck} = userInput;
+
+  const handleUserInput = (key: string) => (text: string) => {
```
화살표 함수를 중첩할수도 있나요?

minju1009 replied:
함수를 리턴하는 함수를 화살표 함수로 쓸 수 있다고 합니다! 그러나 읽는 사람이 불편하기에 수정해주었습니다!

jasonkang14 reviewed (changes requested):
debounce는 잘 구현하셨습니다! 
하지만 `useEffect` hook은 dependency가 있더라도 최초 렌더링 시에 한 번 호출됩니다. 
따라서 지금 구조상으로는 사용자가 이메일을 한 번 입력하기 전에도 api를 호출할 것 같습니다. 
email 값이 입력되지 않은 경우에는 호출하지 않도록 수정해주세요!

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,178 @@
+      body: JSON.stringify({
+        last_name: lastName,
+        first_name: firstName,
+        email: email,
```
ES6 문법에서는 key와  value의 값이 동일하면 생략할 수 있습니다
```typescript
{
  last_name: lastName,
  first_name: firstName,
  email,
  password
}
```

minju1009 replied:
key:value값이 같을떄는 생략해줘도 되는군요! 추후에는 위와 같이 작성하도록 하겠습니다.

minju1009 reviewed (commented):
`useRef`를 활용하여 최초 렌더링 시에는  debounce통신이 일어나지 않도록 처리해주었습니다! 
또한 기존에 계속 할당이 되던 값들을 `memoization` 해 주었으며 추가로 눈 아이콘을 클릭시 패스워드가 보였다 안보였다 하도록 기능 추가해주었습니다.

minju1009 committed:
FIx: Signup 기능 수정사항 반영

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,220 @@
+  };
+
+  useEffect(() => {
+    if (initialRender.current) {
```
`useRef()`를 사용해서 `render`를 확인하는 것도 좋지만, 
지금 상황은 이메일이 빈값인지 아닌지를 파악하는 것이 더 정확할 것 같습니다.

minju1009 replied:
네, 이메일이 빈값인지 아닌지에 따라 API를 호출하도록 수정하였습니다.

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,220 @@
+    return password === passwordCheck;
+  }, [password, passwordCheck]);
+
+  useEffect(() => {
```
여기도 `useMemo()`를 사용하면 state를 하나 줄일 수 있습니다.

minju1009 committed:
Fix : 수정사항 반영
- 1. 'email'이 없으면을useEffect  조건으로 수정
- 2. 'useMemo()'를 통해 isEverythingValid를 state에서 변수로 수정

minju1009 commented:
Comment마다 댓글을 달았었는데, 이상하게 오늘보니 사라져있네요^^; 
주신 피드백 다시 반영하였습니다 :) 피드백 주셔서 감사합니다!

minju1009 commented:
### PR Type
- [ ] 버그 수정
- [ ] 신규 기능
- [x] 기타 (기타인 경우 내용 입력 :  피드백 반영 수정) 

### 해당 PR의 목적
- 회원가입 기능 수정

### 해당 PR 사유
- 회원가입 기능 피드백 사항 수정
    - 1. 'email'이 공란이 아니면 API를 호출하는 것으로 조건 변경
    - 2. 'useMemo()'를 통해 isEverythingValid를 state에서 변수로 수정

jasonkang14 commented on src/screens/Signup/index.tsx:
```diff
@@ -1,55 +1,210 @@
+  };
+
+  useEffect(() => {
+    if (!email) {
```
if 인 경우 리턴하니 else 는 굳이 작성 안하셔도 될 것 같습니다. 성향 차이고 개발팀 차이지만 저희는 그렇게 하고 있어요!

고생하셨습니다 ~

minju1009 replied:
넵! 알려주셔서 감사합니다 :) 굳이 불필요한 코드는 줄여보도록 하겠습니다!

jasonkang14 reviewed (approved):
LGTM
